File: A5_MissedConnections/missed/MissedConnection.java
package missed;

import java.io.IOException;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.ArrayList;
import java.util.List;

import org.apache.hadoop.conf.Configured;
import org.apache.hadoop.fs.Path;
import org.apache.hadoop.io.Text;
import org.apache.hadoop.mapreduce.Job;
import org.apache.hadoop.mapreduce.Mapper;
import org.apache.hadoop.mapreduce.Reducer;
import org.apache.hadoop.mapreduce.lib.input.FileInputFormat;
import org.apache.hadoop.mapreduce.lib.output.FileOutputFormat;
import org.apache.hadoop.util.Tool;
import org.apache.hadoop.util.ToolRunner;

import missed.FlightUtils;

/**
 * Missed Connection -- The map reduce program that computes 
 * the number of missed connections and the percentage of missed connections
 * per airline, per year.
 * @author Harshali Singh, Vishal Mehta
 */
public class MissedConnection extends Configured implements Tool {

	/**
	 * Mapper class to read the entire data and filter out the flights which
	 * do not pass sanity test and output appropriate Key, Value pairs.
	 * Key is the Carrier code, Year and the origin or destination. 
	 * Value is FlightDate, Scheduled Departure Time, Actual Departure Time
	 * and Cancelled value. 
	 */

	public static class ConnectionMapper extends Mapper<Object, Text, Text, Text> {

		/**
		 *	Map Method to process and emit the key value pair described above to the reducer.
		 *  @param   Object  key
		 *  @param   Text    value 
		 *  @return  void
		 */
		public void map(Object key, Text value, Context context) throws IOException, InterruptedException {

			//Split the record
			String line = value.toString();
			String newLine = line.replaceAll("\"", "");
			String formattedLine = newLine.replaceAll(", ", ":");
			String[] row = formattedLine.split(",");

			//Check if the flight passed sanity test.
			if (FlightUtils.sanityTest(row)) {

				String carrierCode = row[6];
				String year = row[0];
				String flightDate = row[5];

				String origin = row[14];
				String dest = row[23];

				//Scheduled Departure and Arrival Time
				String schDepTime = row[29];
				String schArrTime = row[40];

				//Actual Departure and Arrival Time
				String actDepTime = row[30];
				String actArrTime = row[41];

				String cancelled = row[47];

				Text fOutKey = null;
				Text fOutValue = null;
				Text gOutKey = null;
				Text gOutValue = null;

				fOutKey = createKey(carrierCode, year, dest);
				fOutValue = arrCreateValue(flightDate, schArrTime, actArrTime, cancelled);

				gOutKey = createKey(carrierCode, year, origin);
				gOutValue = depCreateValue(flightDate, schDepTime, actDepTime, cancelled);

				//F data
				context.write(fOutKey, fOutValue);

				//G data
				context.write(gOutKey, gOutValue);

			}

		}

		/** Method to Create Mapper output Key
		 *  @param   String    carrier code, year, airport code
		 *  @return  String    Mapper output Key
		 */
		private Text createKey(String carrier, String year, String airport) {
			Text returnKey = null;
			if (!carrier.isEmpty() && !year.isEmpty() && !airport.isEmpty()) {
				returnKey = new Text(carrier.trim() + "\t" + year.trim() + "\t" + airport.trim());
			}
			return returnKey;
		}

		/** Method to Create Mapper output Value
		 *  @param    flight date, scheduled arrival time, actual arrival time, flight is cancelled or not
		 *  @return   Mapper output value for arriving flight
		 */
		private Text arrCreateValue(String flightDate, String schArrTime, String actArrTime, String cancelled) {
			Text returnValue  = new Text("F" + "\t" + flightDate + "\t" + schArrTime + "\t" + actArrTime + "\t" + cancelled);

			return returnValue;
		}

		/** Method to Create Mapper output Value
		 *  @param    flight date, scheduled arrival time, actual arrival time, flight is cancelled or not
		 *  @return   Mapper output value for departing flight
		 */
		private Text depCreateValue(String flightDate, String schDepTime, String actDepTime, String cancelled) {
			Text returnValue  = new Text("G" + "\t" + flightDate + "\t" + schDepTime + "\t" + actDepTime + "\t" + cancelled);

			return returnValue;
		}


	}

	/**
	 * Reducer class to emit the number of connections and number of missed connections 
	 * per carrier, per year and per airport hop connection. The <Key, List<Values>> are for carriers
	 * which have  F.Destination = G.Origin 
	 * 
	 */
	public static class ConnectionReducer extends Reducer<Text, Text, Text, Text>{

		/**
		*	Reduce Method to process the input from mapper and emit the key value pairs.
		* 	Key is Carrier code, year and origin/destination.
		*	Value is total number of connections and number of missed connections.
		*  @param   Text 		   key
		*  @param   Iterable<Text> values 
		*  @return  void
		*/
		public void reduce(Text key, Iterable<Text> values, Context context)
				throws IOException, InterruptedException, IndexOutOfBoundsException {

			SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd");

			int numCons = 0;
			int missedCons = 0;
			
			List<Text> outValues = new ArrayList<Text>();
			List<Text> inValues = new ArrayList<Text>();
			
			String f = new String("F");   //Incoming Flight
			String g = new String("G");   //Outgoing Flight

			for (Text value: values){

				String[] split = value.toString().split("\t");

				if(split[0].trim().equals(f)){

					outValues.add(new Text(value));

				}
				
				if(split[0].trim().equals(g)){

					inValues.add(new Text(value));
				}

			}
			


			for(Text outValue : outValues){

				try{

					String[] outsplit = outValue.toString().split("\t");

						Date fdate = sdf.parse(outsplit[1]);   //Flight Date
						String schArr = outsplit[2];
						String actArr = outsplit[3];
						int fCancelled = Integer.parseInt(outsplit[4]);

						for(Text inValue :  inValues){

							String[] insplit = inValue.toString().split("\t");

								Date gdate = sdf.parse(insplit[1]);         //Flight Date
								String schDep = insplit[2];
								String actDep = insplit[3];

								long diffDays = getDiffDays(fdate, gdate);       //Difference in Days

								//Flights on same day
								if(diffDays == 0){

									int SchDepTime = FlightUtils.convertTime(schDep);
									int ActDepTime = FlightUtils.convertTime(actDep);

									int SchArrTime = FlightUtils.convertTime(schArr);
									int ActArrTime = FlightUtils.convertTime(actArr);

									int schLayover = getLayover(SchDepTime, SchArrTime);
									int actLayover = getLayover(ActDepTime, ActArrTime);

									// Is a connection if scheduled departure of G is <= 6 hours 
									// and >= 30 minutes after the scheduled arrival of F.
									if(schLayover <= 360 && schLayover >= 30){

										numCons += 1;

										// Is a missed connection if Flight F is cancelled 
										// or actual layover is less than 30 minutes 
										if((fCancelled == 1) || actLayover < 30){

											missedCons += 1;

										}


									}								

								} 
								//Flights rolled over to next day
								else if(diffDays == 1 && gdate.after(fdate)){


									String strSchDep = FlightUtils.appendString(schDep);
									String strActDep = FlightUtils.appendString(actDep);

									String strSchArr = FlightUtils.appendString(schArr);
									String strActArr = FlightUtils.appendString(actArr);

									//Scheduled Departure
									int schDepHour = getHour(strSchDep);
									int schDepMin = getMin(strSchDep);

									//Scheduled Arrival
									int schArrHour = getHour(strSchArr);
									int schArrMin = getMin(strSchArr);		

									//Actual Departure
									int actDepHour = getHour(strActDep);
									int actDepMin = getMin(strActDep);

									//Actual Arrival
									int actArrHour = getHour(strActArr);
									int actArrMin = getMin(strActArr);							

									//Scheduled Layover
									int schLayover = _getLayover(schArrHour, schArrMin, schDepHour, schDepMin);

									//Actual Layover
									int actLayover = _getLayover(actArrHour, actArrMin, actDepHour, actDepMin);

									//Connection
									if(schLayover <= 360 && schLayover >= 30){

										numCons += 1;

										//Missed Connection
										if((fCancelled == 1) || (actLayover < 30)){

											missedCons += 1;
											
										
										}


									}



								}

							}
						}
				
				catch(ParseException ex){

				}



			}

			context.write(key, new Text(numCons + "\t" + missedCons));


		}
		/**
		*	Method to calculate difference in days between arriving and departing flights.
		*  @param   Date fdate
		*  @param   Date gdate 
		*  @return  long diffdays (difference in days)
		*/

		private static long getDiffDays(Date fdate, Date gdate){

			long diff = Math.abs(gdate.getTime() - fdate.getTime());  //Date difference in milliseconds
			long diffDays = diff / (24 * 60 * 60 * 1000);             //in Days

			return diffDays;

		}

		/**
		*	Method to extract hour from the received time.
		*  @param   String strTime
		*  @return  int    hour 
		*/

		private static int getHour(String strTime){

			return strTime.length() == 4? Integer.parseInt(strTime.substring(0, 2)) : 0;
		}

		/**
		*	Method to extract minutes from the received time.
		*  @param   String strTime
		*  @return  int    minutes 
		*/

		private static int getMin(String strTime){

			return strTime.length() == 4? Integer.parseInt(strTime.substring(2, 4)) : 0;
		}

		/**
		*	Method to calculate layover if flights are on same day.
		*  @param   int depTime
		*  @param	int arrTime
		*  @return  int difference between departure and arrival time.  
		*/
		private static int getLayover(int depTime, int arrTime){

			return Math.abs(depTime - arrTime);

		}

		
		/**
		*	Method to calculate layover if flights are NOT on same day.
		*  @param   int arrHour, arrMin, depHour, depMin
		*  @return  int layover in minutes. 
		*/
		private static int _getLayover(int arrHour, int arrMin, int depHour, int depMin){

			int _arrHour = Math.abs(24-arrHour);
			int _hour = _arrHour + depHour;
			int _min = Math.abs(depMin - arrMin);
			int _layover = _hour * 60 - _min;

			return _layover;

		}

	}

	/**
	 *	Driver Program to run the jobs and set the input and output paths.
	 */
	public int run (String[] args) throws Exception {

		Job job = Job.getInstance();

		job.setJar("MissedConnection.jar");
		job.setMapperClass(ConnectionMapper.class);
		job.setReducerClass(ConnectionReducer.class);
		
		job.setMapOutputKeyClass(Text.class);
		job.setMapOutputValueClass(Text.class);

		job.setOutputKeyClass(Text.class);
		job.setOutputValueClass(Text.class);

		FileInputFormat.addInputPath(job, new Path(args[0]));
		FileOutputFormat.setOutputPath(job, new Path(args[1]));

		return job.waitForCompletion(true) ? 0 : 1;
	}

	public static void main(String[] args) throws Exception {

		System.exit(ToolRunner.run(new MissedConnection(), args));
	}
}

File: A6_Scala/MissedConnection.scala
//Author: Harshali Singh, Vishal Mehta

import org.apache.spark.SparkContext
import org.apache.spark.SparkContext._
import org.apache.spark.SparkConf
import org.apache.spark.rdd.RDD
import java.util.ArrayList

import org.joda.time.DateTime
import org.joda.time.format.DateTimeFormat
import org.joda.time.Days

/**
 * Missed Connection -- The scala program that computes 
 * the number of missed connections and the percentage of missed connections
 * per airline, per year.
 */

object MissedConnection{

	def main(args: Array[String]){

	if (args.length < 2) {

		System.err.println("Usage: MissedConnection <input> <output>")
		System.exit(1)
		}

		val inputPath = args(0)
		val outPath = args(1)

		val sparkConf = new SparkConf().
		setAppName("Missed Connection").
		setMaster("local")
		val sc = new SparkContext(sparkConf)
		val t0 = System.nanoTime()

		//Read csv files
		val csv = sc.textFile(inputPath)

		//Format and Trim lines
		val formatted = csv.map(line => line.replaceAll("\"", "").replaceAll(", ", ":"))
		val rows = formatted.map(line => line.split(",").map(_.trim))
		val header = rows.first()

		//Remove Header
		val data = rows.filter(line => line(0) != header(0))

		//Filter data based on Sanity Test
		val saneData = data.filter(line => sanityTest(line))

		//Arriving Flights: (Carrier Year AiportCode, Array(Flight Date, Scheduled Arrival, Actual Arrival, Cancelled))
		val fRDD = saneData.map(

			row => (row(6)+"\t"+row(0)+"\t"+row(23), Array(row(5), row(40), row(41), row(47)))

			)

		//Departing Flights: (Carrier Year AiportCode, Array(Flight Date, Scheduled Departure, Actual Departure))
		val gRDD = saneData.map(

			row => (row(6)+"\t"+row(0)+"\t"+row(14), Array(row(5), row(29), row(30)))

			)

		//Performed Co-Group on F and G RDD
		val newRDD = (fRDD cogroup gRDD)

		val missedRDD = newRDD.map(key => {

			val (k, v) = key;
			val (f, g) = v;

			var cons = 0
			var missedcons = 0

			val format = DateTimeFormat.forPattern("yyyy-MM-dd");

			//Array lists to store F and G data 
			val foutvalues = new ArrayList[String];
			val ginvalues = new ArrayList[String];

			//Map on F Data 
			val fdata = f.map (fentry => {
				
				val fDate = fentry(0);

				//Arrival Times
				var schArr = appendString(fentry(1));
				var actArr = appendString(fentry(2));

				//Cancelled
				var fCancelled = fentry(3);

				foutvalues.add(fDate + "\t" + schArr + "\t" + actArr + "\t" + fCancelled)

				})
			
			//Map on G Data 
			val gdata = g.map (gentry => {
				
				val gDate = gentry(0);

				//Departure Times
				var schDep = appendString(gentry(1));
				var actDep = appendString(gentry(2));

				ginvalues.add(gDate + "\t" + schDep + "\t" + actDep)

				})
			
			val fsize = foutvalues.size
			val gsize = ginvalues.size

			//Iterating over F data Arraylist 
			for(i <- 0 until fsize) {				

				val fsplit = foutvalues.get(i).split('\t')

				val strFdate = fsplit(0)
				val fDate = format.parseDateTime(strFdate)

				val schArr = fsplit(1)
				val actArr = fsplit(2)
				val fCancelled = fsplit(3).toInt

					//Iterating over G data Arraylist 
					for (j <- 0 until gsize) {

						val gsplit = ginvalues.get(j).split('\t')

						val strGdate = gsplit(0)
						val gDate = format.parseDateTime(strGdate)

						val schDep = gsplit(1)
						val actDep = gsplit(2)

						//Flights on same day
						if(getDiffDays(fDate, gDate) == 0) {

							if(isSameDayCon(schDep, schArr)) {

								cons += 1
								
								// if flight on same day is missed
								if(isSameDayMissedCon(actDep, actArr, fCancelled)) {

									missedcons += 1

								} 
							}
							
						}
						
						//Flights rolled over to Next day
						else if (getDiffDays(fDate, gDate) == 1 && gDate.isAfter(fDate)) {

							if(isNextDayCon(schDep, schArr)) {

									cons += 1

								// if flight on different day is missed
								if (isNextDayMissedCon(actDep, actArr, fCancelled)) {

									missedcons += 1
										
								} 
								
							} 

						}						

					}
			}			
			
			//Emit a tuple containing Key and number of connections and missed connections
			(k, (cons.toString,missedcons.toString))
	
		}).collect()


		//Create Output RDD[String]
	    val output = missedRDD.map (value => {

	   		val (k,v) = value;
	   		val (p1,p2) = v;

	   		k + "\t" + p1 + "\t" +	 p2

	   	})

	    //Save Result to Disk
	   	val result = sc.parallelize(output)
        result.saveAsTextFile(outPath)
        
    	// Shut down Spark, avoid errors.
    	sc.stop()

    	// Total running time of program
    	val t1 = System.nanoTime()
    	println("Total time:======>" + ((t1 - t0)/1000000) + " ms")

    }

   /**
	* Method to check if given two times on different days, the flight is a connection or not
	* @param    String        scheduled arrival and departure times of two flights
	* @return   Boolean       value depending on whether the flight is a connection or not
	*
	**/

    def isNextDayCon(schDep : String, schArr : String) : Boolean = {

    	var strSchDep = appendString(schDep);
    	var strSchArr = appendString(schArr);

		//Scheduled Departure
		var schDepHour = getHour(strSchDep);
		var schDepMin = getMin(strSchDep);

		//Scheduled Arrival
		var schArrHour = getHour(strSchArr);
		var schArrMin = getMin(strSchArr);								

		//Scheduled Layover
		var schLayover = _getLayover(schArrHour, schArrMin, schDepHour, schDepMin);

		if(schLayover <= 360 && schLayover >= 30){

			return true;
		}

		return false;

	}

   /**
	* Method to check if given two times on different days, the flight is a missed connection or not
	* @param    String        actual arrival and departure times of two flights
	* @return   Boolean       value depending on whether the flight is a missed connection or not
	*
	**/
	def isNextDayMissedCon(actDep : String, actArr : String, fCancelled : Int) : Boolean = {

		var strActDep = appendString(actDep);
		var strActArr = appendString(actArr);

		//Actual Departure
		var actDepHour = getHour(strActDep);
		var actDepMin = getMin(strActDep);

		//Actual Arrival
		var actArrHour = getHour(strActArr);
		var actArrMin = getMin(strActArr);	

		//Actual Layover
		var actLayover = _getLayover(actArrHour, actArrMin, actDepHour, actDepMin);

		// Is a missed connection if Flight F is cancelled 
		// or actual layover is less than 30 minutes
		if((fCancelled == 1) || actLayover < 30){

			return true;
		}

		return false;
	}

   /**
	* Method to check if given two times on same day, the flight is a connection or not
	* @param    String        scheduled arrival and departure times of two flights
	* @return   Boolean       value depending on whether the flight is a connection or not
	*
	**/
	def isSameDayCon(schDep : String, schArr : String) : Boolean = {

		var schDepTime = convertTime(schDep);
		var schArrTime = convertTime(schArr);
		var schLayover = getLayover(schDepTime, schArrTime);

		if(schLayover <= 360 && schLayover >= 30){

			return true;
		}

		return false;

	}

   /**
	* Method to check if given two times on same day, the flight is a missed connection or not
	* @param    String        actual arrival and departure times of two flights
	* @return   Boolean       value depending on whether the flight is a missed connection or not
	*
	**/

	def isSameDayMissedCon(actDep : String, actArr : String, fCancelled : Int) : Boolean = {

		var actDepTime = convertTime(actDep);
		var actArrTime = convertTime(actArr);

		//Actual Layover
		var actLayover = getLayover(actDepTime, actArrTime);
		
		// Is a missed connection if Flight F is cancelled 
		// or actual layover is less than 30 minutes
		if((fCancelled == 1) || actLayover < 30){

			return true;
		}

		return false;

	}


   /**
	* Method to get the number of days between two dates
	* @param    DateTime      scheduled departure flight dates of two flights
	* @return   Int           number of days between the two dates
	*
	**/
	def getDiffDays(start : DateTime, end : DateTime) : Int = {

		var diffDays = Days.daysBetween(start.toLocalDate(), end.toLocalDate()).getDays()   //In Days

		return diffDays

	}

   /**
	* Method to get the Hour of a given time in HHmm format
	* @param    String      Time in hhmm format
	* @return   Int         the hour part
	*
	**/
	def getHour(strTime : String) : Int = {

		if(strTime.length() == 4){
			return (strTime.substring(0, 2)).toInt
			} else return 0;

		}

   /**
	* Method to get the Minute of a given time in HHmm format
	* @param    String      Time in hhmm format
	* @return   Int         the minute part 
	*
	**/
	def getMin(strTime : String) : Int = {

		if(strTime.length() == 4){
			return (strTime.substring(2, 4)).toInt
			} else return 0;
	}

   /**
	* Method to get the layover time between two given flight times on same day
	* @param    Int      Time in minutes
	* @return   Int      Layover time 
	*
	**/
	def getLayover(depTime : Int, arrTime : Int) : Int = {

		return math.abs(depTime - arrTime);

	}

   /**
	* Method to get the layover time between two given flight times on different days
	* @param    Int      Time in minutes
	* @return   Int      Layover time 
	*
	**/
	def _getLayover(arrHour : Int, arrMin : Int, depHour : Int, depMin : Int) : Int = {

		var _arrHour = math.abs(24-arrHour);
		var _hour = _arrHour + depHour;
		var _min = math.abs(depMin - arrMin);
		var _layover = _hour * 60 - _min;

		return _layover;

	}


   /**
	* Method to enforce the sanity test for every row (flight data).
	* @param    Array[String] each row of flight data
	* @return   Boolean       value depending on whether a flight passes sanity test or not.
	*
	**/
	def sanityTest (row : Array[String]) : Boolean = {

		if (row.length != 110){
			return false;
		}

		try {

				//Time is converted to minutes
				var CRSArrTime = if (row(40).isEmpty()) 0 else convertTime(row(40));
				var CRSDepTime = if (row(29).isEmpty()) 0 else convertTime(row(29));
				var CRSElapsedTime = (row(50)).toFloat.toInt;
				var timeZone = CRSArrTime - CRSDepTime - CRSElapsedTime;
				
				//CRSArrTime and CRSDepTime should not be zero
				
				if (CRSArrTime == 0 || CRSDepTime == 0) {
					return false;
				}	
				
				//timeZone % 60 should be 0
				if (timeZone % 60 != 0) {
					return false;
				}
				
				var OriginAirportId = (row(11)).toFloat.toInt;
				var DestAirportId = (row(20)).toFloat.toInt;
				var OriginAirportSeqId = (row(12)).toFloat.toInt;
				var DestAirportSeqId = (row(21)).toFloat.toInt;
				var OriginCityMarketId = (row(13)).toFloat.toInt;
				var DestCityMarketId = (row(22)).toFloat.toInt;
				var OriginStateFips = (row(17)).toFloat.toInt;
				var DestStateFips = (row(26)).toFloat.toInt;
				var OriginWac = (row(19)).toFloat.toInt;
				var DestWac = (row(28)).toFloat.toInt;
				
				//AirportID,  AirportSeqID, CityMarketID, StateFips, Wac should be larger than 0
				if (OriginAirportId <= 0 || DestAirportId <= 0 || OriginAirportSeqId <= 0 || 
					DestAirportSeqId <= 0 || OriginCityMarketId <= 0 || DestCityMarketId <= 0 || 
					OriginStateFips <= 0 || DestStateFips <= 0 || OriginWac <= 0 || DestWac <= 0) {
					return false;
				}
				
				//Origin, Destination,  CityName, State, StateName should not be empty
				if (row(14).isEmpty() || row(23).isEmpty() || row(15).isEmpty() || row(24).isEmpty() ||
					row(16).isEmpty() || row(25).isEmpty() || row(18).isEmpty() || row(27).isEmpty()) {
					return false;
				}
				
				//Flights that are not Cancelled
				var cancelled = (row(47)).toFloat.toInt;
				if (cancelled != 1) {
					var arrTime = if(row(41).isEmpty()) 0 else convertTime(row(41));
					var depTime = if(row(30).isEmpty()) 0 else convertTime(row(30));
					var actualElapsedTime = (row(51)).toFloat.toInt;
					
					//ArrTime -  DepTime - ActualElapsedTime - timeZone should be zero
					var diff = arrTime - depTime - actualElapsedTime - timeZone;
					if (diff != 0 && diff % 1440 != 0) {
						return false;
					}
					
					var arrDelay = (row(42)).toFloat.toInt;
					var arrDelayMinutes = (row(43)).toFloat.toInt;
					
					//if ArrDelay > 0 then ArrDelay should equal to ArrDelayMinutes
					//if ArrDelay < 0 then ArrDelayMinutes should be zero
					if (arrDelay > 0 && arrDelay != arrDelayMinutes) {
						return false;
						} else if (arrDelay < 0 && arrDelayMinutes != 0) {
							return false;
						}

					//if ArrDelayMinutes >= 15 then ArrDel15 should be true
					if (arrDelayMinutes >= 15 && (row(44)).toFloat.toInt != 1) {
						return false;
					}
				}
				
			}
			catch { case ex: NumberFormatException => {
				return false;
			}	
		}

		return true;	
	}

   /**
	* Method to convert time in hh:mm format to Minutes.
	* @param String s (time in hh:mm)
	* @return int    time in minutes
	*/
	def convertTime(s : String) : Int = {

		var result = appendString(s);
		var time = 0;

		if(result.length() == 4){
			var hours = (result.substring(0,2)).toInt;
			var minutes = (result.substring(2,4)).toInt;

			time = hours * 60 + minutes;

		}

		return time;

	}

   /**
	* Method to append number of zeros to the time based on its length
	* @param String s (time in hh:mm)
	* @return String appended string with zeros.
	*/
	def appendString(s : String) : String = {

		var result = "";
		var s_new = s.trim();
		if(s_new.length()< 4 && !s_new.isEmpty()){

			result = "0" + s_new;
		}
		else if (s_new.isEmpty()) {

			result = "0000"

		} else {
			
			result = s_new
		}

		return result;
	}
}

